runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcGetDataStream.java

```diff
@@ -342,64 +379,92 @@ private void queueRequestAndWait(QueuedRequest request) throws InterruptedExcept
       batch.addRequest(request);
     }
     if (responsibleForSend) {
-      if (waitForSendLatch == null) {
+      if (prevBatch == null) {
         // If there was not a previous batch wait a little while to improve
         // batching.
-        Thread.sleep(1);
+        sleeper.sleep(1);
       } else {
-        waitForSendLatch.await();
+        prevBatch.waitForSendOrFailNotification();
       }
       // Finalize the batch so that no additional requests will be added.  Leave the batch in the
       // queue so that a subsequent batch will wait for its completion.
       synchronized (batches) {
-        verify(batch == batches.peekFirst());
+        verify(batch == batches.peekFirst(), "GetDataStream request batch removed before send().");
         batch.markFinalized();
       }
-      sendBatch(batch.requests());
+      trySendBatch(batch);
+    } else {
+      // Wait for this batch to be sent before parsing the response.
+      batch.waitForSendOrFailNotification();
+    }
+  }
+
+  void trySendBatch(QueuedBatch batch) {
+    try {
+      sendBatch(batch.sortedRequestsReadOnly());
       synchronized (batches) {
```

Comment:
changing all `synchronized (batches)` blocks to be like the following should fix mutations to `batches` and shutdown racing

```
synchronized (shutdownLock) {
  if (isShutdown()) {
    // cleanup anything that needs to be cleaned up;
    return;
    // returning here should be safe as shutDown cleared all existing `batches`
  }
  synchronized (batches) {
  ...
  }
}
}
```

Maybe we could use  only `shutdownLock` to guard access to `batches` and remove `synchronized (batches)`, will there be any downsides in doing so?

Reply:
done

we can do that since we batches isn't used to guard anything else